tools/Dashboard/DevHome.Dashboard/Services/WidgetHostingService.cs

```diff
@@ -15,25 +17,121 @@ public class WidgetHostingService : IWidgetHostingService
+            {
+                return await Task.Run(() => _widgetHost.GetWidgets());
+            }
+            catch (COMException ex) when (ex.HResult == RpcServerUnavailable || ex.HResult == RpcCallFailed)
+            {
+                _log.Warning(ex, $"Failed to operate on out-of-proc object with error code: 0x{ex.HResult:x}");
+                _widgetHost = null;
+
+                // If the host died, the catalog probably did too.
+                _widgetCatalog = null;
+            }
+            catch (Exception ex)
+            {
+                _log.Error(ex, "Exception getting widgets from service:");
+            }
+        }
+
+        // If we lost the object, create a new one. This call will get the WidgetService back up and running.
         if (_widgetHost == null)
         {
             try
             {
                 _widgetHost = await Task.Run(() => WidgetHost.Register(new WidgetHostContext("BAA93438-9B07-4554-AD09-7ACCD7D4F031")));
+                return await Task.Run(() => _widgetHost.GetWidgets());
             }
             catch (Exception ex)
             {
-                _log.Error(ex, "Exception in WidgetHost.Register:");
+                _log.Error(ex, "Exception getting widgets from service:");
             }
         }
 
-        return _widgetHost;
+        return null;
     }
 
+    /// <summary>
+    /// Create and return a new widget.
+    /// </summary>
+    /// <returns>The new widget, or null if one could ont be created.</returns>
+    public async Task<Widget> CreateWidgetAsync(string widgetDefinitionId, WidgetSize widgetSize)
+    {
+        // If we already have a WidgetHost, check if the COM object is still alive and use it if it is.
+        if (_widgetHost != null)
+        {
+            try
+            {
+                return await Task.Run(async () => await _widgetHost.CreateWidgetAsync(widgetDefinitionId, widgetSize));
+            }
+            catch (COMException ex) when (ex.HResult == RpcServerUnavailable || ex.HResult == RpcCallFailed)
+            {
+                _log.Warning(ex, $"Failed to operate on out-of-proc object with error code: 0x{ex.HResult:x}");
+                _widgetHost = null;
+
+                // If the host died, the catalog probably did too.
+                _widgetCatalog = null;
+            }
+            catch (Exception ex)
+            {
+                _log.Error(ex, "Exception creating a widget:");
+            }
+        }
+
+        if (_widgetHost == null)
+        {
+            try
+            {
+                _widgetHost = await Task.Run(() => WidgetHost.Register(new WidgetHostContext("BAA93438-9B07-4554-AD09-7ACCD7D4F031")));
+                return await Task.Run(async () => await _widgetHost.CreateWidgetAsync(widgetDefinitionId, widgetSize));
+            }
+            catch (Exception ex)
+            {
+                _log.Error(ex, "Exception creating a widget:");
+            }
+        }
+
+        return null;

```

Comment:
Suggestion with no context about widgets 🙂 

```suggestion
  var attempt = 0;
   while(attempt++ < MaxAttempt)
   {
            try
            {
               _widgetHost ??= await Task.Run(() => WidgetHost.Register(new WidgetHostContext("BAA93438-9B07-4554-AD09-7ACCD7D4F031")));
                return await Task.Run(async () => await _widgetHost.CreateWidgetAsync(widgetDefinitionId, widgetSize));
            }
            catch (COMException ex) when (ex.HResult == RpcServerUnavailable || ex.HResult == RpcCallFailed)
            {
                _log.Warning(ex, $"Failed to operate on out-of-proc object with error code: 0x{ex.HResult:x}");
                _widgetHost = null;
                _widgetCatalog = null;
            }
            catch (Exception ex)
            {
                _log.Error(ex, "Exception creating a widget:");
            }
        }
   }
     
   return null;
```

Reply:
Since this code is repeated, we can also promote it to a method and pass a lambda as argument for the core logic.
Example [WinGetRecovery.cs](https://github.com/microsoft/devhome/blob/main/tools/SetupFlow/DevHome.SetupFlow/Services/WinGet/WinGetRecovery.cs)

Reply:
I love this simpler version. I'm going to keep the repeated code (not the lambda) for this change.